### DISCUSSION TITLE
Accurate location - city level data

### DISCUSSION THREAD

**Discussion by metmarkosaric** [plausible/analytics#102] [Anything Else]
"It would be great to get a little more info about the location. Not too much to invade privacy but I’m part of a small business and it’s hard to get valuable info when page visits come just out of Switzerland and Germany.

Agreed, I don’t think capturing the user city is a privacy issue and it would be useful to see.

I totally agree with this feature. To get more information as it now is like country. It should give information about country, state and city.That is important to make some better and more precise Settings when you use google ads.

I’d assume the IPv4 and IPv6 IP addresses are anonymized though it doesn’t provide anonymity to some companies companies due to their IP address ranges."

**Update:** Cities and regions are now live!

**Comment by rhysperry111** [comment]
if this feature does get implemented, would it be possible to turn it off? I do not need data that specific (although I understand that others might) so in the best interests of privacy I don't want to collect more data than I need

**Reply by vectorselector** [reply]
It would seem that simply adding support for the `GeoLite2-city` database to the Map on Plausible will require you to add/enable access for this database, and thereby satisfy your wishes.
We gain license for `GeoLite2-city` as well as `GeoLite2-country` when signing up for Max,Mind databases.
For USA, for example, not having at least state-by-state breakdowns is a non-starter. City is arguably a fine substitute, but more granularity than nation is required, not optional, for any serious use-case. This is not personally identifiable information!
The IP is anonymized and we are talking about NYC, for example.

**Comment by flatroof** [comment]
Small businesses/local would need far less geo than country. For US, state and city minimal.

**Comment by erikjoling** [comment]
I would also like to see more specific location data. This would make Plausible much more interesting for small businesses whose audiences are more more local, regional or national. And it would make it easier for me to convince them to not use Google Analytics :)

**Comment by danesnick** [comment]
Would love to have this feature, at minimum at state level basis, especially since I am a US user. Is there any plans for this to be on the roadmap in the near future?

**Reply by ukutaht** [reply]
@danesnick Sorry no definitive plans at the moment. I think this is important but we have things like API, import/export etc. that are higher in priority

**Reply by macmcmeans** [reply]
I respect that other items on your roadmap come before this one, but geo tracking that includes only country simply isn't useful. From a novelty standpoint it may be interesting, but from an I-need-to-know-where-my-visits-are-coming-from perspective city/state geo is a necessity, not an option.

There will be plenty of people, like myself, willing to start the 30-day demo in order to kick the tires to see if it's possible to kick GA to the curb. Upon discovering that your granularity extends no farther than country level tracking, you can expect there will be plenty of people, like myself, who will have to forego your alternative until you address this deficiency. 

I can wait till tomorrow for the API, and the import/export functionality to be rolled out, but I need meaningful geo tracking now. Respectfully, I urge you to please consider revising your roadmap for us.

**Reply by jakezien** [reply]
I started my 30-day trial yesterday, and was excited to check my dashboard today to see the data. I discovered I'm in exactly the same boat as @nicholasdanes and @macmcmeans: US user, expecting mostly US visitors, finding the coarseness of the geo data so limiting that Plausible is unlikely to be a complete-enough solution for me to subscribe. Which is a bummer, because I love everything else about the product. 

I agree completely with the post above: I can wait on an API, but more granular location data is of immediate, critical value. I know firsthand that nothing in product development is easy.  It would seem, though, that if you're already using IP to determine country, then querying for cities or states/provinces/territories should be a simple enough change. Even a plaintext table of this data would make Plausible useful enough to buy; no need for a whole map UI. 

I, too, respectfully urge you to reconsider the priority of this addition. I think **for most users, more granular location data is table stakes for any analytics product.** Without it, the product feels quite incomplete — so for those of us in this thread, this is less a feature request than a product requirement.

**Reply by macmcmeans** [reply]
100% agreed!

**Comment by jmatsuba** [comment]
I'm evaluating self hosted options for web analytics as one of our clients has data storage requirements. Plausible is a great fit but not having better geolocation data is a deal breaker. 

So +1 for this feature request.

I'd be game to throw some side projects hours at this, but would want someone to work on it with. I don't have a geolocation or an Elixir background, so ramp up would be a bit hard I suspect.  However, sounds like fun!

**Reply by metmarkosaric** [reply]
nice to hear @jmatsuba, thanks! this is definitely one of those highly requested features and we'd like to add it. would be great to get some help so we can do it faster. 

someone mentioned we could use https://db-ip.com/db/lite.php as an alternative geolocation database which includes cities too. there's the #219 thread where you can learn more.

**Comment by jakezien** [comment]
Bumping this further to note it was the #3 most upvoted request on the old roadmap.

**Reply by metmarkosaric** [reply]
thanks @jakezien! this is definitely one of those highly requested features and we'd like to add it. we're a small team with limited development resources so cannot promise any timelines at this time on when it will be completed. stay tuned!

**Comment by IzakJackson** [comment]
This is also something I would love to see. My business mainly operates in the UK, and this would make tracking marketing efforts targetted at different counties/cities much easier!

**Comment by LinusGeffarth** [comment]
Is this anywhere near implementation?
And will historic data show the more detailed region info or will that only work for new data?

**Reply by metmarkosaric** [reply]
All our development is done in the open. We haven't started on this work yet and I cannot promise any specifics on when this will be done. At the moment, our focus is on completing the embed mode and on releasing the first stable version of the stats API. We're a small team with limited resources so need to focus on things one step at a time. If someone from the community contributes with this, it will obviously speed things up.

**Reply by ukutaht** [reply]
To add, it will only work for new data. We don't store IP addresses so we cannot generate location data for historical stats.

**Reply by LinusGeffarth** [reply]
All good, was just curious :)

**Comment by jakezien** [comment]
Awesome to see a version of embeddable mode ship today! Cheers to the team — very exciting for your community to see consistent progress.

You know where this is going: any chance you'll show some love next for more granular location data? It's got more votes than embeddable mode on this Discussions board. My trial ends in two weeks, and I'd really, really love to pay you at the end of it. 🙏

**Reply by metmarkosaric** [reply]
thanks @jakezien! All our development is done in the open. We haven't started on this work yet and I cannot promise any specifics on when this will be done. At the moment, our focus is on releasing the first stable version of the stats API, then doing some performance updates and then starting the work on the sites API. We're a small team with limited resources so need to focus on things one step at a time. If someone from the community contributes with this, it will obviously speed things up.

**Comment by omarius** [comment]
For my project, it is absolutely necessary to get more details than the country only. At least the state the visits are coming from should be viewed.

**Comment by bradkane** [comment]
Just wanted to let everyone know that @aymanterra and I have started working on adding city level data to Plausible.  Here are some initial screenshots:

![image](https://user-images.githubusercontent.com/42099127/123302173-6d880a80-d4ea-11eb-9d17-09f96ca7e335.png)
![image](https://user-images.githubusercontent.com/42099127/123302275-8abcd900-d4ea-11eb-8461-c82b93f636ce.png)

Mapping the data creates some interesting challenges as current tool does not seem to support province/city details.  Any ideas @ukutaht, @metmarkosaric or anyone else?

BTW, we are using masked IP addresses as this seems to be forgone conclusion in the near future.

**Reply by metmarkosaric** [reply]
Sounds great @bradkane and @aymanterra, looking forward to playing around with this! 

What's your thinking on whether we can make this work with the current country drilldown? The map view doesn't seem to be too useful on the drilldown. Not sure how easy it would be to replace the map with a list of states or cities in the drilldown view? So when I'm filtering by a country, the report there automatically on the main view gives me top cities or states. We could also add the tabs on the report so people can switch from list of states to list of cities or regions/provinces similar like it can be done on the Devices report now.

Thanks for all your efforts on this!

**Reply by bradkane** [reply]
Interesting ideas on tabs and drilldowns.  Right now the hierarchy screenshots are triggered by the "Details" link.  What about adding multi-select filtering to Plausible?  This is particularly important with geographic tracking as the user might want to select multiple countries (e.g. EU countries, or US + Canada) and potentially multiple provinces/states or cities.  Currently, only a single selection from a category can be made.  If we do this, we should consider extending the multi-select filters to all categories (e.g. devices, browsers, etc.)

Currently, there is no map detail below country outlines.  To show a concentration of page visits visually, we are thinking of using a heat map alternative driven by longitude/latitude data instead of a mapping API which would allow zooming, etc.  Have you considered any APIs to go below country level visuals?  Our thinking is that someone who is aware of his/her own country would be able to see the heat concentration and know the city(ies) or provinces where the greatest traffic occurred.  Phase 1 might include this more limited visualization of visits, and in a later release, we could try to implement a full mapping API (like Google Maps, MapQuest, etc.).

**Reply by aodesky** [reply]
This looks perfect. Let me know if there's any way I can help. I have some basic familiarity with html.

**Comment by metmarkosaric** [comment]
Cities and regions are now finally live on Plausible following a great demand!

We've been testing a few different geolocation databases over the last few months and settled on https://db-ip.com. It's closest to what we want to achieve but we're still not happy with the way they display some cities in some countries. In some cases, they go to the district/borough level and we'd prefer all visits to be on the city level only (for instance by default DB-IP would display Soho or King's Cross rather than London on the city level).

We've done some manual overrides ourselves for some cities and plan to do more as we discover them. If you find these situations for the cities you're familiar with, please add them [to this thread](https://github.com/plausible/analytics/issues/1582) and we'll take a look in order to manually override them. Thanks!

And thanks to @bradkane and @aymanterra for all the help on this!

**Reply by brambow** [reply]
Great news! Can you add regions/cities to the CSV exports as well?

**Reply by metmarkosaric** [reply]
thanks! this is already being worked on. you can follow the progress here: https://github.com/plausible/analytics/pull/1485

**Reply by aodesky** [reply]
Thanks for your efforts! Will this feature be released to the plausible/hosting repo anytime soon?

**Reply by metmarkosaric** [reply]
thanks! last self hosted release was last month so next one should hopefully be around May this year

**Reply by varac** [reply]
Great, really looking forward to next release. If I would build a container image from latest master now, would that include the city feature ?

**Reply by varac** [reply]
> We've been testing a few different geolocation databases over the last few months and settled on https://db-ip.com. It's closest to what we want to achieve but we're still not happy with the way they display some cities in some countries. In some cases, they go to the district/borough level and we'd prefer all visits to be on the city level only (for instance by default DB-IP would display Soho or King's Cross rather than London on the city level).

@metmarkosaric Indeed, the accuracy of db-ip.com is far from ideal. When I visit db-ip's main website it tells me I'm visiting from Frankfurt while I'm appox. 500km away. I guess Frankfurt is displayed as a country fallback when db-ip can't resolve better. So it's not a coincidence most visitors of the plausible website as show incorrectly as visiting from Frankfurt as I noticed on https://plausible.io/plausible.io?period=6mo: 

![image](https://user-images.githubusercontent.com/488213/161245948-3ce6d584-42a0-4fab-83d2-32aab3aa11c2.png)

Personally I changed to [Maxmind's GeoLite2 City DB](https://dev.maxmind.com/geoip/geolite2-free-geolocation-data?lang=en) which is much more accurate, at least it resolved my city correctly in contrast to db-ip. Downside is that you need to register first, but the download is free of charge.

**Reply by aodesky** [reply]
> thanks! last self hosted release was last month so next one should hopefully be around May this year

Great! Just wanted to follow up on this. Will the update to the plausible/hosting repo be happening sometime soon?

**Reply by ukutaht** [reply]
Yes. Next week

**Reply by marc22761** [reply]
> > We've been testing a few different geolocation databases over the last few months and settled on https://db-ip.com. It's closest to what we want to achieve but we're still not happy with the way they display some cities in some countries. In some cases, they go to the district/borough level and we'd prefer all visits to be on the city level only (for instance by default DB-IP would display Soho or King's Cross rather than London on the city level).
> 
> @metmarkosaric Indeed, the accuracy of db-ip.com is far from ideal. When I visit db-ip's main website it tells me I'm visiting from Frankfurt while I'm appox. 500km away. I guess Frankfurt is displayed as a country fallback when db-ip can't resolve better. So it's not a coincidence most visitors of the plausible website as show incorrectly as visiting from Frankfurt as I noticed on https://plausible.io/plausible.io?period=6mo:
> 
> ![image](https://user-images.githubusercontent.com/488213/161245948-3ce6d584-42a0-4fab-83d2-32aab3aa11c2.png)
> 
> Personally I changed to [Maxmind's GeoLite2 City DB](https://dev.maxmind.com/geoip/geolite2-free-geolocation-data?lang=en) which is much more accurate, at least it resolved my city correctly in contrast to db-ip. Downside is that you need to register first, but the download is free of charge.

I'm having the same issue. 95% of my visitors are from Frankfurt. I'm probably a good portion of this and I'm 500km away from Frankfurt. Any chance we get a better resolution on the hosted product as well?

**Reply by LinusGeffarth** [reply]
I second this

**Reply by metmarkosaric** [reply]
Thanks for the feedback @marc22761! 

You can test your IP addresses on their website to see what location they show: https://db-ip.com

Here's what they say on their website: "We aggregate data from a large number of sources and through agreements with popular ISPs. We constantly evolve our machine learning models for more accurate IP address geolocation." 

Unfortunately, when DB-IP gives us the wrong city for some IP address, there's nothing we can do on our end to fix it but we can hope that their database improves. We constantly get any improvements that they make so when they make updates, our geolocation accuracy improves too.

We have tested other databases including Maxmind but for instance Maxmind doesn't give any location at all for vast majority of visitors so we can only display the location of a small share of visitors when using Maxmind.

**Reply by marc22761** [reply]
@metmarkosaric Thanks for responding so quickly. Support is top notch and I'm not complaining / blaming about the inaccuracy. I just have to asses if plausible is meeting the clients expectation.

The reason why its showing hamburg is probably, that Frankfurt is the node where the vodafone cable network hooks up with the internet. dp-ip is showing:

`{
  "ipAddress": "37.4.254.XX",
  "continentCode": "EU",
  "continentName": "Europe",
  "countryCode": "DE",
  "countryName": "Deutschland",
  "isEuMember": true,
  "currencyCode": "EUR",
  "currencyName": "Euro",
  "phonePrefix": "49",
  "languages": [
    "de"
  ],
  "stateProvCode": "HE",
  "stateProv": "Hessen",
  "district": "Darmstadt",
  "city": "Frankfurt am Main",
  "geonameId": 2925533,
  "zipCode": "60313",
  "latitude": 50.1109,
  "longitude": 8.68213,
  "gmtOffset": 2,
  "timeZone": "Europe/Berlin",
  "weatherCode": "GMXX0040",
  "asNumber": 3209,
  "asName": "VODANET",
  "isp": "Vodafone Kabel Deutschland",
  "usageType": "corporate",
  "organization": "Kabel Deutschland Customer Services",
  "isCrawler": false,
  "isProxy": false,
  "threatLevel": "low"
}`

**Reply by marc22761** [reply]
But how come that other services like NordVPN are more accurate? BTW: Its still a 6km kilometers away, the postcode is not accurate, but its the right city not 500km away

<img width="1503" alt="Bildschirmfoto 2022-05-09 um 10 13 31" src="https://user-images.githubusercontent.com/105198922/167368478-b9f04064-215a-4de9-8f5e-2e3ec8c56391.png">

**Reply by metmarkosaric** [reply]
they must be using a different database provider. i couldn't find any info on their site about which one though

**Reply by marc22761** [reply]
@metmarkosaric I know. The only problem is, I cannot tell the client that it is not possible, if he can find something that is proving this wrong.

I'm sorry to say, but I see zero chance that the client will switch if we cant do better than this.

**Reply by metmarkosaric** [reply]
sure, no problem. that makes sense. we cannot find a better ip database that works for web analytics at the moment unfortunately. nordvpn is not an analytics product and these web based ip checkers have different needs than analytics products do so may be able to use different databases and other tools too. we'll keep an eye on different new options that come to the market but db-ip seems to be the strongest one at the moment from our testing.

**Reply by metmarkosaric** [reply]
hi @marc22761 and @LinusGeffarth! we were able to start a dialogue around this with DB-IP and they've agreed to manually review any suggestions/improvements that we send them in order to update their database. the IP address range shared in this thread `37.4.254.XX` has been handed to their data team, and after they've manually confirmed it, they will update their databases and services which will then make its way to Plausible. if you do get any more reports of inaccuracies please do share them with us and we will reach out to DB-IP for them to review. thanks!

**Reply by bradkane** [reply]
Hi @ukutaht @metmarkosaric - @aymanterra  and I are anxiously awaiting the pending update on self-hosting so we can implement city-level tracking.  Anything holding this up that we can help with?

**Reply by ukutaht** [reply]
The release is delayed because I am looking to delegate more responsibilities, self-hosted releases included. The plan is for @cnkk to be responsible for future releases but he is currently busy with some other stuff. I cannot give an estimate at the moment, I'm sorry about the delay.

**Reply by bradkane** [reply]
Understood.  We will rearrange our plans accordingly.

**Comment by varac** [comment]
@metmarkosaric Could you please pass on the `84.158.175.xx` block to the db-ip data team ?
It's still resolving falsely at db-ip.com:

```
❯ curl http://api.db-ip.com/v2/free/84.158.175.2
{
    "ipAddress": "84.158.175.2",
    "continentCode": "EU",
    "continentName": "Europe",
    "countryCode": "DE",
    "countryName": "Germany",
    "stateProv": "Hesse",
    "city": "Frankfurt am Main"
}
```

while it's correct at i.e. ipinfo.io:

```
❯ curl -s ipinfo.io/84.158.175.2  
{
  "ip": "84.158.175.2",
  "hostname": "p549eaf02.dip0.t-ipconnect.de",
  "city": "Hamburg",
  "region": "Hamburg",
  "country": "DE",
  "loc": "53.5518,9.9415",
  "org": "AS3320 Deutsche Telekom AG",
  "postal": "22767",
  "timezone": "Europe/Berlin",
  "readme": "https://ipinfo.io/missingauth"
}
```
Thanks!

**Reply by metmarkosaric** [reply]
thanks @varac! i've shared it with the DB-IP team

**Comment by NathanC** [comment]
I just integrated with Plausible (self hosting it), and I don't see any city data, only counties. Does it take time to aggregate city data, or does the self-hosting not provide it?

**Reply by metmarkosaric** [reply]
Our self-hosted version is not updated as frequently as the Plausible managed hosting in the cloud, which changes daily. This means that the latest features won't be immediately available. Think of self-hosted as a long-term release. We stack all the new features and fixes that have been battle-tested on the hosted version and release a new self-hosted version twice a year or so. We don't have any more specific timelines than this as we have limited resources, and the self-hosted release depends on the free capacity of our developers.

**Reply by aodesky** [reply]
Is there actually a plan to add city-level data to the self-hosted version? Although it keeps getting hinted at in the comments, the indefinite wait suggests this is being treated as a paid-only feature.

**Reply by metmarkosaric** [reply]
@aodesky there hasn't been a new version of self hosted code released since we launched regions/cities which is why it's not available in the self hosted release yet. should be there in the next version

**Reply by Matb85** [reply]
Hi, could you disclose when Cities/Region will be available on the self-hosted version? I switched from GA in October, and I don't know if/when I should update. 

Keep up the good work! Your solution is way better privacy-wise than GA.

**Comment by metmarkosaric** [comment]
Update: after some more testing of different geolocation databases, we have now gone live with a switch of the database provider. we now use the MaxMind database instead. no database is 100% accurate unfortunately but this change should provide some better accuracy in terms of cities. details https://plausible.io/docs/countries#how-it-works
